File: data/build-shade.js
#!/usr/bin/env node
// Builds 4 merged shade GeoJSON files (one per time snapshot) for one-shot loading.
// Output: frontend/data/shade_1000.geojson
//         frontend/data/shade_1300.geojson
//         frontend/data/shade_1530.geojson
//         frontend/data/shade_1800.geojson
//
// Strategy:
//   1. Douglas-Peucker at DP_EPSILON removes redundant vertices.
//   2. Rings with fewer than MIN_AREA_DEG2 area are dropped (sub-pixel at zoom 13).
//   3. Coordinates are rounded to 5 decimal places (~1 m).
//   4. Only the shade value for that time slot is kept (property "s", integer).
// This produces one compact file per slot that the browser loads once at startup.

const fs   = require("fs");
const path = require("path");

const IN_DIR  = path.join(__dirname, "raw", "shade");
const OUT_DIR = path.join(__dirname, "..", "frontend", "data");

const SHADE_SLOTS = [
  { key: "1000", prop: "shade_percent_at_1000" },
  { key: "1300", prop: "shade_percent_at_1300" },
  { key: "1530", prop: "shade_percent_at_1530" },
  { key: "1800", prop: "shade_percent_at_1800" },
];

const DP_EPSILON   = 0.00005;  // ~5 m — sub-pixel detail removed at zoom 14
const MIN_AREA     = 5e-9;     // drop rings with area < ~0.5 m² (genuine noise)

// ── Helpers ───────────────────────────────────────────────────────────────────
function dpDist(px, py, ax, ay, bx, by) {
  const dx = bx - ax, dy = by - ay;
  if (dx === 0 && dy === 0) return Math.hypot(px - ax, py - ay);
  const t = Math.max(0, Math.min(1, ((px - ax) * dx + (py - ay) * dy) / (dx * dx + dy * dy)));
  return Math.hypot(px - ax - t * dx, py - ay - t * dy);
}

function dpSimplify(pts, eps) {
  if (pts.length <= 2) return pts;
  let maxD = 0, idx = 0;
  const last = pts.length - 1;
  for (let i = 1; i < last; i++) {
    const d = dpDist(pts[i][0], pts[i][1], pts[0][0], pts[0][1], pts[last][0], pts[last][1]);
    if (d > maxD) { maxD = d; idx = i; }
  }
  if (maxD > eps) {
    const a = dpSimplify(pts.slice(0, idx + 1), eps);
    const b = dpSimplify(pts.slice(idx), eps);
    return [...a.slice(0, -1), ...b];
  }
  return [pts[0], pts[last]];
}

// Shoelace area (absolute value)
function ringArea(ring) {
  let a = 0;
  for (let i = 0, j = ring.length - 1; i < ring.length; j = i++) {
    a += (ring[j][0] + ring[i][0]) * (ring[j][1] - ring[i][1]);
  }
  return Math.abs(a) / 2;
}

function roundCoord([lng, lat]) {
  return [Math.round(lng * 1e5) / 1e5, Math.round(lat * 1e5) / 1e5];
}

function simplifyRing(ring) {
  const s = dpSimplify(ring, DP_EPSILON).map(roundCoord);
  if (s.length < 4) return null;
  if (s[0][0] !== s[s.length-1][0] || s[0][1] !== s[s.length-1][1]) s.push(s[0]);
  if (ringArea(s) < MIN_AREA) return null;
  return s;
}

function simplifyGeom(geom) {
  if (geom.type === "Polygon") {
    const rings = geom.coordinates.map(simplifyRing).filter(Boolean);
    return rings.length ? { type: "Polygon", coordinates: rings } : null;
  }
  if (geom.type === "MultiPolygon") {
    const polys = geom.coordinates
      .map(poly => { const rings = poly.map(simplifyRing).filter(Boolean); return rings.length ? rings : null; })
      .filter(Boolean);
    return polys.length ? { type: "MultiPolygon", coordinates: polys } : null;
  }
  return geom;
}

// ── Load & simplify once (geometry is shared across all 4 slots) ─────────────
const files = fs.readdirSync(IN_DIR).filter(f => f.endsWith(".geojson"));
console.log(`Reading ${files.length} neighbourhood files…`);

const allRaw = [];
for (const file of files) {
  const gj = JSON.parse(fs.readFileSync(path.join(IN_DIR, file), "utf8"));
  for (const feat of gj.features) {
    if (feat.geometry) allRaw.push(feat);
  }
}
console.log(`Total raw features: ${allRaw.length}`);

console.log("Simplifying…");
const simplified = [];
for (const feat of allRaw) {
  const geom = simplifyGeom(feat.geometry);
  if (geom) simplified.push({ geom, props: feat.properties });
}
console.log(`Retained: ${simplified.length} features (${(simplified.length/allRaw.length*100).toFixed(1)}%)`);

// ── Write one file per time slot ─────────────────────────────────────────────
for (const { key, prop } of SHADE_SLOTS) {
  const features = simplified.map(({ geom, props }) => ({
    type: "Feature",
    properties: { s: Math.round(props[prop] ?? 0) },
    geometry: geom,
  }));

  const outPath = path.join(OUT_DIR, `shade_${key}.geojson`);
  fs.writeFileSync(outPath, JSON.stringify({ type: "FeatureCollection", features }));

  const kb = fs.statSync(outPath).size / 1024;
  console.log(`shade_${key}.geojson — ${features.length} features, ${kb.toFixed(0)} KB`);
}

File: data/build-shade.py
#!/usr/bin/env python3
"""Rebuild the sidewalk-shade layer from the SlimShady source.

The previous files were misshapen because they were simplified with a ~5 m
Douglas-Peucker epsilon — larger than the 2-4 m width of a sidewalk, which
collapses the polygons.

This rebuild:
  * keeps shape — Douglas-Peucker at EPS (~0.4 m), only removing genuinely
    redundant near-collinear vertices that the GIS export densified;
  * stores the geometry ONCE in a single file, with the shade percentage for
    every slot kept as integer properties s1000/s1300/s1530/s1800 (0-100).
    (Shipping four copies of the same geometry was ~4x larger for no reason.)
  * rounds coordinates to 6 decimals (~0.11 m).

Usage:  python3 data/build-shade.py
Source: ../SlimShady/data/Buurt_data/*.geojson  (Gebruiksfunctie == "Sidewalk")
Output: data/shade.geojson
"""
import glob, json, math, os

HERE    = os.path.dirname(os.path.abspath(__file__))
SRC_DIR = os.path.join(HERE, "..", "..", "SlimShady", "data", "Buurt_data")
SLOTS   = [("s1000", "shade_percent_at_1000"),
           ("s1300", "shade_percent_at_1300"),
           ("s1530", "shade_percent_at_1530"),
           ("s1800", "shade_percent_at_1800")]
OUT     = os.path.join(HERE, "shade.geojson")
EPS     = 4e-6   # ~0.4 m — well below sidewalk width, preserves shape
NDIGITS = 6


def dp_simplify(pts, eps):
    """Iterative Douglas-Peucker; keeps the first/last point fixed."""
    n = len(pts)
    if n <= 2:
        return pts
    keep = [False] * n
    keep[0] = keep[-1] = True
    stack = [(0, n - 1)]
    while stack:
        a, b = stack.pop()
        ax, ay = pts[a]; bx, by = pts[b]
        dx, dy = bx - ax, by - ay
        seg2 = dx * dx + dy * dy
        maxd, idx = eps, -1
        for i in range(a + 1, b):
            px, py = pts[i]
            if seg2 == 0:
                d = math.hypot(px - ax, py - ay)
            else:
                t = ((px - ax) * dx + (py - ay) * dy) / seg2
                t = 0.0 if t < 0 else 1.0 if t > 1 else t
                d = math.hypot(px - ax - t * dx, py - ay - t * dy)
            if d > maxd:
                maxd, idx = d, i
        if idx != -1:
            keep[idx] = True
            stack.append((a, idx)); stack.append((idx, b))
    return [pts[i] for i in range(n) if keep[i]]


def clean_ring(ring):
    simplified = dp_simplify(ring, EPS)
    out = []
    for x, y in simplified:
        rx, ry = round(x, NDIGITS), round(y, NDIGITS)
        if not out or out[-1][0] != rx or out[-1][1] != ry:
            out.append([rx, ry])
    if len(out) >= 2 and out[0] != out[-1]:
        out.append(out[0])
    return out if len(out) >= 4 else None


def clean_polygon(rings):
    cleaned = [r for r in (clean_ring(r) for r in rings) if r]
    return cleaned or None


def iter_polys(geom):
    if geom["type"] == "Polygon":
        p = clean_polygon(geom["coordinates"])
        if p:
            yield p
    elif geom["type"] == "MultiPolygon":
        for poly in geom["coordinates"]:
            p = clean_polygon(poly)
            if p:
                yield p


def pct(props, prop):
    v = props.get(prop)
    return 0 if v is None else max(0, min(100, int(round(v))))


def main():
    files = sorted(glob.glob(os.path.join(SRC_DIR, "*.geojson")))
    if not files:
        raise SystemExit(f"No source files found in {SRC_DIR}")
    print(f"Reading {len(files)} source files")

    features = []
    n_src = 0
    for fp in files:
        with open(fp) as fh:
            fc = json.load(fh)
        for feat in fc.get("features", []):
            props = feat.get("properties", {})
            if props.get("Gebruiksfunctie") != "Sidewalk":
                continue
            geom = feat.get("geometry")
            if not geom:
                continue
            n_src += 1
            sprops = {k: pct(props, p) for k, p in SLOTS}
            for poly in iter_polys(geom):
                features.append({
                    "type": "Feature",
                    "properties": sprops,
                    "geometry": {"type": "Polygon", "coordinates": poly},
                })

    with open(OUT, "w") as fh:
        json.dump({"type": "FeatureCollection", "features": features}, fh,
                  separators=(",", ":"))
    mb = os.path.getsize(OUT) / 1e6
    print(f"Sidewalks: {n_src} read -> {len(features)} polygons")
    print(f"Wrote {os.path.basename(OUT)}: {mb:.1f} MB")


if __name__ == "__main__":
    main()
